### DISCUSSION TITLE
When ever trying to draw with a paint in which shader is custom sksl, SEHException is thrown

### DISCUSSION THREAD

**Discussion by martinrhan** [AvaloniaUI/Avalonia#16039] [Q&A]
I made a minimal example that repeats the bug. This is my MainView.cs
```
public partial class MainView : UserControl {
    public MainView() {
        InitializeComponent();
        Dispatcher.UIThread.Post(
            () => {
                using (DrawingContext context = ViewModel.Bitmap.CreateDrawingContext()) {
                    context.Custom(new CustomDrawOperation());
                }
                Image.InvalidateVisual();
            }, 
            DispatcherPriority.ApplicationIdle);
    }

    MainViewModel ViewModel => (MainViewModel)DataContext!;

    public override void EndInit() {
        base.EndInit();
    }

    public class CustomDrawOperation : ICustomDrawOperation {
        static CustomDrawOperation() {
            string sksl = """
                    half4 main(vec2 fragCoord) { 
                        return vec4(0,0,0,255);
                    }
                    """;
            string error;
            var runtimeEffect = SKRuntimeEffect.Create(sksl, out error);
            var shader = runtimeEffect.ToShader(false);
            paint = new SKPaint {
                Shader = shader,
            };
        }

        private static SKPaint paint;

        public Rect Bounds { get; }

        public void Dispose() {
        }

        public bool Equals(ICustomDrawOperation? other) {
            return ReferenceEquals(this, other);
        }

        public bool HitTest(Point p) {
            return Bounds.Contains(p);
        }

        public void Render(ImmediateDrawingContext context) {
            var skiaFeature = context.TryGetFeature(typeof(ISkiaSharpApiLeaseFeature)) as ISkiaSharpApiLeaseFeature;
            if (skiaFeature == null) { return; }
            using (var lease = skiaFeature.Lease()) {
                SKCanvas canvas = lease.SkCanvas;
                canvas.DrawPaint(paint); \\ SEHException occurs here
            }
        }
    }
```
The error code of SEHException is -2147467259, which is 0x80004005. In [Common HRESULT Values](https://learn.microsoft.com/en-us/windows/win32/seccrypto/common-hresult-values), it's E_FAIL, Unspecified failure. I have no idea what's happening.
Please download the zip file for more detail.
[AvaloniaApplication2.zip](https://github.com/user-attachments/files/15856409/AvaloniaApplication2.zip)

**Comment by rabbitism** [comment]
IIRC sksl only works in Skiasharp 3 beta. There is a new method `SKRuntimeEffect.CreateShader`
(Avalonia 11.1 should be compatible with Skiasharp 3, you can manually reference the package in your project. )

**Reply by martinrhan** [reply]
It can work in version 11.0.10. Strangely, after changing it to calling `context.Custom(new CustomDrawOperation())` in `public override void Render(DrawingContext context)` of MainView and removing the Image control, it works

**Reply by martinrhan** [reply]
That means currently custom sksl shader drawing can work in drawing context given in overrided Render function of visuals, but not in RenderTargetBitmap.

**Reply by timunie** [reply]
Please try 11.1 beta and probably Skia 3.0 beta. Would be interesting if that would solve the issue or not

**Reply by martinrhan** [reply]
For exactly same code, 11.0.2 and 11.0.10 worked. Then I changed it to 11.1.0-rc1, no exception but it didn't work (the MainView is white instead of grey). By placing breakpoints, I found that `MainView.Render` (which contains `context.Custom(new CustomDrawOperation());`) was called, but `CustomDrawOperation.Render` was not called. This is very strange and did not happen in older version.

**Reply by martinrhan** [reply]
I tried 11.1.0-rc1 for original version of demo app (rending to the RenderTargetBitmap), the SEHException bug is present.
